File: app/src/main/java/eu/warble/voice/util/aStarSearch/AStarSearch.java
package eu.warble.voice.util.aStarSearch;

import com.indoorway.android.common.sdk.model.IndoorwayNode;

import java.util.*;

/**
 * The AStarSearch class, along with the AStarNode class,
 * implements a generic A* search algorithm. The AStarNode
 * class should be subclassed to provide searching capability.
 */
public class AStarSearch {

    List<IndoorwayNode> nodeList;

    public final List<IndoorwayNode> findPath(IndoorwayNode startNode, IndoorwayNode endNode) {

        AStarNodeWrapper startWrapper = new AStarNodeWrapper(startNode);
        List<AStarNodeWrapper> openList = new LinkedList<>();
        List<AStarNodeWrapper> closedList = new LinkedList<>();
        openList.add(new AStarNodeWrapper(startNode)); // add starting node to open list

        Boolean done = false;
        AStarNodeWrapper current;
        while (!done) {
            current = lowestFInOpen(openList); // get node with lowest fCosts from openList
            closedList.add(current); // add current node to closed list
            openList.remove(current); // delete current node from open list

            if ((current.getIndoorwayNode().equals(endNode))) { // found goal
                return calcPath(startWrapper, current);
            }


            // for all adjacent nodes:
            List<AStarNodeWrapper> adjacentNodes = getAdjacent(current);
            for (int i = 0; i < adjacentNodes.size(); i++) {
                AStarNodeWrapper currentAdj = adjacentNodes.get(i);
                if (!openList.contains(currentAdj)) { // node is not in openList
                    currentAdj.setPrevious(current); // set current node as previous for this node
                    currentAdj.sethCosts(endNode); // set h costs of this node (estimated costs to goal)
                    currentAdj.setgCosts(current); // set g costs of this node (costs from start to this node)
                    openList.add(currentAdj); // add node to openList
                } else { // node is in openList
                    if (currentAdj.getgCosts() > currentAdj.calculategCosts(current)) { // costs from current node are cheaper than previous costs
                        currentAdj.setPrevious(current); // set current node as previous for this node
                        currentAdj.setgCosts(current); // set g costs of this node (costs from start to this node)
                    }
                }
            }

            if (openList.isEmpty()) { // no path exists
                return new LinkedList<IndoorwayNode>(); // return empty list
            }
        }
        return null; // unreachable
    }

    private List<AStarNodeWrapper> getAdjacent(AStarNodeWrapper current) {
        List<AStarNodeWrapper> result = new LinkedList<>();

        Collection<Long> neighbours = current.getIndoorwayNode().getNeighbours();
        for (Long id :
                neighbours) {
            for (IndoorwayNode node :
                    nodeList) {
                if (node.getId() == id) {
                    result.add(new AStarNodeWrapper(node));
                    break;
                }
            }
        }
        return result;
    }

    private AStarNodeWrapper lowestFInOpen(List<AStarNodeWrapper> openList) {
        // TODO currently, this is done by going through the whole openList!
        AStarNodeWrapper cheapest = openList.get(0);
        for (int i = 0; i < openList.size(); i++) {
            if (openList.get(i).getfCosts() < cheapest.getfCosts()) {
                cheapest = openList.get(i);
            }
        }
        return cheapest;
    }

    private List<IndoorwayNode> calcPath(AStarNodeWrapper start, AStarNodeWrapper goal) {
        // TODO if invalid nodes are given (eg cannot find from
        // goal to start, this method will result in an infinite loop!)
        LinkedList<AStarNodeWrapper> path = new LinkedList<>();

        AStarNodeWrapper curr = goal;
        boolean done = false;
        while (!done) {
            path.addFirst(curr);
            curr = (AStarNodeWrapper) curr.getPrevious();

            if (curr.equals(start)) {
                done = true;
            }
        }
        List<IndoorwayNode> indoorwayNodes = new LinkedList<>();
        for (AStarNodeWrapper node :
                path) {
            indoorwayNodes.add(node.getIndoorwayNode());
        }
        return indoorwayNodes;
    }
}

File: app/src/main/java/eu/warble/voice/util/AStarSearch.kt
package eu.warble.voice.util

import com.indoorway.android.common.sdk.model.IndoorwayNode
import eu.warble.voice.data.model.Node
import java.util.*
import kotlin.collections.HashSet


class AStarSearch(list: List<IndoorwayNode>) {
    private var nodeList: LinkedList<Node> = LinkedList()

    init {
        list.forEach {
            nodeList.add(Node(it))
        }
    }

    fun findPath(from: IndoorwayNode, to: IndoorwayNode): List<IndoorwayNode>{
        val fromNode = getNodeWithId(from.id)
        val toNode = getNodeWithId(to.id)
        val explored = HashSet<Node?>()

        val queue: PriorityQueue<Node?> = PriorityQueue(500, { node1: Node?, node2: Node? ->
            when {
                node1 != null && node2 != null -> when {
                    node1.f_scores > node2.f_scores -> return@PriorityQueue 1
                    node1.f_scores < node2.f_scores -> return@PriorityQueue -1
                    else -> return@PriorityQueue 0
                }
                else -> return@PriorityQueue 0
            }
        })

        fromNode?.g_scores = 0.0
        queue.add(fromNode)
        while (!queue.isEmpty()){
            val current = queue.poll()
            explored.add(current)

            if(current!= null && current == toNode) {
                return makePath(toNode)
            }

            if (current != null) {
                for (nodeId in current.indoorwayNode.neighbours){
                    val child = getNodeWithId(nodeId)
                    if (child != null){
                        val cost = child.indoorwayNode.coordinates.getDistanceTo(current.indoorwayNode.coordinates)
                        val temp_g_scores = current.g_scores + cost
                        val temp_f_scores = temp_g_scores + child.h_scores;
                        if (explored.contains(child) && (temp_f_scores >= child.f_scores)){
                            continue
                        }else if(!queue.contains(child) || temp_f_scores < child.f_scores){
                            child.parent = current
                            child.g_scores = temp_g_scores
                            child.f_scores = temp_f_scores
                            if (queue.contains(child)) {
                                queue.remove(child)
                            }
                            queue.add(child)
                        }
                    }
                }
            }
        }
        return emptyList()
    }

    private fun makePath(toNode: Node): List<IndoorwayNode> {
        val result = LinkedList<IndoorwayNode>()

        var node: Node? = toNode
        while (node != null) {
            result.add(node.indoorwayNode)
            node = node.parent
        }
        Collections.reverse(result)
        return result
    }

    private fun getNodeWithId(id: Long): Node? {
        return nodeList.find {
            it.indoorwayNode.id == id
        }
    }
}